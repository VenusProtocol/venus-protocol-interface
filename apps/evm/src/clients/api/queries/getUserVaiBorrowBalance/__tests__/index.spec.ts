import type { PublicClient } from 'viem';

import fakeAccountAddress, {
  altAddress as fakeVaiControllerContractAddress,
} from '__mocks__/models/address';

import { getUserVaiBorrowBalance } from '..';

const fakeVaiRepayAmountWithInterests = 10000000000n;

describe('getUserVaiBorrowBalance', () => {
  it('throws an error when the multicall fails', async () => {
    const multicallMock = vi.fn(() => [
      undefined,
      {
        result: undefined,
        status: 'failure',
      },
    ]);

    const fakePublicClient = {
      multicall: multicallMock,
    } as unknown as PublicClient;

    try {
      await getUserVaiBorrowBalance({
        accountAddress: fakeAccountAddress,
        publicClient: fakePublicClient,
        vaiControllerContractAddress: fakeVaiControllerContractAddress,
      });

      throw new Error('getUserVaiBorrowBalance should have thrown an error but did not');
    } catch (error) {
      expect(error).toMatchInlineSnapshot('[Error: somethingWentWrong]');
    }
  });

  it('returns the VAI fee with interests', async () => {
    const multicallMock = vi.fn(() => [
      undefined,
      {
        result: fakeVaiRepayAmountWithInterests,
        status: 'success',
      },
    ]);

    const fakePublicClient = {
      multicall: multicallMock,
    } as unknown as PublicClient;

    const response = await getUserVaiBorrowBalance({
      accountAddress: fakeAccountAddress,
      publicClient: fakePublicClient,
      vaiControllerContractAddress: fakeVaiControllerContractAddress,
    });

    expect(multicallMock).toHaveBeenCalledTimes(1);
    expect(multicallMock).toHaveBeenCalledWith({
      contracts: [
        {
          abi: expect.any(Object),
          address: fakeVaiControllerContractAddress,
          functionName: 'accrueVAIInterest',
        },
        {
          abi: expect.any(Object),
          address: fakeVaiControllerContractAddress,
          functionName: 'getVAIRepayAmount',
          args: [fakeAccountAddress],
        },
      ],
    });

    expect(response).toMatchSnapshot();
  });
});
