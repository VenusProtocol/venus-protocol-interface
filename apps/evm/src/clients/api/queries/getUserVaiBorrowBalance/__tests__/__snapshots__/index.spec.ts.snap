// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getUserVaiBorrowBalance > returns the VAI fee with interests 1`] = `
{
  "userVaiBorrowBalanceMantissa": "10000000000",
}
`;
